tools/simple_formatter/SimpleFormatterTool.java

```diff
@@ -0,0 +1,199 @@
+  private static final Flag<Boolean> dryRun = Flag.create(false);
+
+  @FlagDesc(name = "java", description = "Format java files")
+  private static final Flag<Boolean> java = Flag.create(false);
+
+  @FlagDesc(name = "proto", description = "Format proto files")
+  private static final Flag<Boolean> proto = Flag.create(false);
+
+  @FlagDesc(name = "python", description = "Format python files")
+  private static final Flag<Boolean> python = Flag.create(false);
+
+  @FlagDesc(name = "cpp", description = "Format C++ (*.cc) files")
+  private static final Flag<Boolean> cpp = Flag.create(false);
+
+  @FlagDesc(name = "ignore_directories", description = "Ignored directories, split by comma")
+  private static final Flag<String> ignoreDirectories = Flag.create("");
+
+  private static boolean isJava(Path file) {
+    return getExtension(file).equals(".java");
+  }
+
+  private static boolean isPython(Path file) {
+    return getExtension(file).equals(".py");
+  }
+
+  private static boolean isProto(Path file) {
+    return getExtension(file).equals(".proto");
+  }
+
+  private static boolean isCpp(Path file) {
+    return getExtension(file).equals(".cc");
+  }
+
+  private static String getExtension(Path file) {
+    String filename = file.getFileName().toString();
+    int dotIndex = filename.lastIndexOf('.');
+    if (dotIndex == -1) {
+      return "";
+    }
+    return filename.substring(dotIndex);
+  }
+
+  private static boolean shouldFormatFile(Path file, Set<Path> ignoredDirectories) {
+    boolean formatByExtension =
+        ((isJava(file) && java.get())
+            || (isProto(file) && proto.get())
+            || (isPython(file) && python.get())
+            || (isCpp(file) && cpp.get()));
+    boolean inIgnoredDirectory =
+        ignoredDirectories.stream().anyMatch(ignoredDir -> file.startsWith(ignoredDir));
+    return formatByExtension && !inIgnoredDirectory;
+  }
+
+  private static Map<String, BaseFormatter> formatters =
+      ImmutableMap.of(
+          ".java", JavaFormatter.init(),
+          ".proto", ClangFormatter.init(),
+          ".py", PythonFormatter.init(),
+          ".cc", ClangFormatter.init());
+
+  public static void main(String[] args) {
+    Flags.parse(args, SimpleFormatterTool.class.getPackage());
+
+    int exit_code = 0;
+    Set<Path> ignoredDirectories =
+        Stream.of(ignoreDirectories.get().split(","))
+            .map(path -> Paths.get(path))
+            .collect(Collectors.toSet());
+
+    try (Stream<Path> stream = Files.walk(Paths.get(path.get()).toAbsolutePath())) {
+      Iterable<Path> paths = () -> stream.filter(Files::isRegularFile).iterator();
+
+      for (Path path : paths) {
+        try {
+          if (shouldFormatFile(path, ignoredDirectories)) {
+            if (dryRun.get()) {
+              System.out.println(
+                  String.format("Intending to format: %s", path.toAbsolutePath().toString()));
+              exit_code = 1;
```

Comment:
Why do we return `exit_code = 1`? and shouldn't we then also return 1 when formatting changed the file?

Reply:
@oferb actually, this is a leftover. I first thought that `--dry-run` in my implementation would print out files that would be **changed** by formatting, but in fact it prints out what files it _intends_ to format (without considering it _needs_ formatting). Since then, I've taken a different  approach: to check, we **do** perform the formatting and then check if `git` reports that something changed.